[Tang Zong Hao] iP

ZHTang29 opened the pull request:
### Duke

> Duke can easily help you free your mind of having to remember a long intensive list of things to do! It is

- text-based
- easy to pick up
- ~~FAST~~ _SUPER FAST_ to use!

All you need to do is, 
1. download it from [here](https://github.com/ZHTang29/ip)
2. double-click it
3. add your tasks
4. let Duke automatically manage your tasks for you! 😄 

Features:
- [X] Adding and deleting tasks
- [X] Finding tasks by name
- [ ] More exciting features coming soon! 👍 

If you are an avid java programmer, you can use it to practice Java too! Here is the `main` method:

```ruby
    public static void main(String[] args) {
        Duke duke = new Duke("tasks.txt");
        duke.run();
    }
```

Commit:
Add Gradle support

ZHTang29 committed:
Level-1

ZHTang29 committed:
Level-2

ZHTang29 committed:
Added ability to mark tasks as undone/done

ZHTang29 committed:
Added functionality of different tasks (ToDo, Deadline, Event)

ZHTang29 committed:
Add testcases for automated text UI testing

ZHTang29 committed:
Added functionality to deal with different errors in User Inputs

ZHTang29 committed:
Added functionality for deleting tasks

ZHTang29 committed:
Added new testcases to test exception handling & new 'delete tasks' features

SeeuSim commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,20 @@
+    @Override
+    public String toString() {
+        String[] strArr = deadline.split(" ", 2);
+        return name + " (" + strArr[0] + ":" + " " + strArr[1] + ")";
```
You may wish to leave a line break or use `String::format` to improve readability:

```java
return String.format(
    "%s (: %s)",
    name,
    strArr[0],
    strArr[1]
);
```

Also, `strArr` should use more declarative naming in the plural form. `tokens` would be a better variable name.

SeeuSim commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,104 @@
+        return "Hello! I'm Duke" + "\n" + "What can I do for you?";
+    }
+
+    public String farewellUser() {
```
More declarative method naming should be used. This method merely generates a farewell message, so `getExitMessage` would be clearer.

Likewise for `greetUser` - it can be clearer if it is named `getWelcomeMessage`

SeeuSim commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,104 @@
-        System.out.println("Hello from\n" + logo);
+        Duke duke = new Duke();
+        TaskList tl = new TaskList();
+        String greeting = duke.greetUser();
```
You may insert line breaks to improve the readability of your code, so that it is clearer what each block is doing, as opposed to a chain of random function calls lumped together.

SeeuSim commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,104 @@
+            try {
+                if (!userInput.equals("list")) {
+                    if (userInput.startsWith("mark")) {
+                        String[] strArr = userInput.split(" ", 2);
```
`tokens` would be preferred over `strArr`

SeeuSim commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,104 @@
+
+        while(!userInput.equals("bye")) {
+            try {
+                if (!userInput.equals("list")) {
```
This sequence of `if-else` blocks is hard to read and understand. A `switch` case based off the first word of the command would be neater.

SeeuSim commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,104 @@
+                        String[] strArr = userInput.split(" ", 2);
+                        String taskName = strArr[1];
+                        ToDo toDoTask = new ToDo(taskName);
+                        String addConfirmation = tl.addTask(toDoTask);
```
`confirmAddMessage` would be a better name

SeeuSim commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,104 @@
+                        Deadline deadlineTask = new Deadline(taskName, dueTime);
+                        String addConfirmation = tl.addTask(deadlineTask);
+                        System.out.println(addConfirmation);
+                        userInput = sc.nextLine();
```
It seems like this code is repeated for each block. You may wish to abstract each block into classes so that it is neater and easier to read.

SeeuSim commented on src/main/java/DukeException.java:
```diff
@@ -0,0 +1,13 @@
+
+    @Override
+    public String toString() {
+        return this.getExceptionType() + " detected! " + super.getMessage();
```
A neater way would be to use `String::format`, so that you declaratively show the format of the output `String`.

```java
return String.format(
    "%s detected! %s", ...
);
```

SeeuSim commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,23 @@
+
+    @Override
+    public String toString() {
+        String[] strArr = from.split(" ", 2);
```
Better variable name would be `fromTokens`. Likewise, `untilTokens`

SeeuSim commented on src/main/java/IncorrectArgumentsException.java:
```diff
@@ -0,0 +1,12 @@
+public class IncorrectArgumentsException extends DukeException {
+
+    public IncorrectArgumentsException(String command, int numberExpected, int numberActual) {
+        super("Sorry, command: " + command + " expected " + numberExpected + " arguments, detected " + numberActual
```
This is rather hard to read. Using `String::format` would be neater:

```java
super(
    String.format(
        "Sorry, command: %s expected %s arguments, detected %s arguments!",
        ...
    )
);
```

SeeuSim commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,31 @@
+
+    public String markAsDone() {
+        this.isDone = true;
+        return "Nice! I've marked this task as done:" + "\n" + "[" + this.getTaskType() + "][X] " + this.toString();
```
Adding line breaks would be neater:

```java
return "Nice! I've marked this task as done:\n"
    + String.format("[%s][X] %s",
            this.getTaskType(),
            ...
        );
```

SeeuSim commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,36 @@
+import java.util.List;
+
+public class TaskList {
+    protected List<Task> list;
```
A List holds many items, so a better name would be `tasks`

SeeuSim commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,36 @@
+    public String addTask(Task task) {
+        list.add(task);
+        items++;
+        return "Got it. I've added this task:" + "\n" + " [" + task.getTaskType() + "][ ] " + task + "\n"
```
Usage of line breaks and `String::format` would be neater

SeeuSim commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,36 @@
+        StringBuilder tasklist = new StringBuilder();
+        tasklist.append("Here are the tasks in your list:" + "\n");
+        for(int i = 0; i < items; i++) {
+            tasklist.append(i + 1).append(".").append("[").append(list.get(i).getTaskType()).append("][")
```
It would be neater to use a StringBuilder or `String::format`

SeeuSim reviewed (commented):
Overall, the coding standard/style seems okay. There's just a few glaring issues:

- Incessant use of `if-else` blocks makes it hard to debug
- String addition lumped together on one line makes it hard to understand the returned String's structure
- Unclear variable names
- Repeated code throughout multiple blocks decreasing readability (solvable with OOP)

Please improve on the above so that I and others can understand your code better 😊 Thank you!

engenhui1999 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,104 @@
+        }
+
+        String farewell = duke.farewellUser();
+        System.out.println(farewell);
```
I think you should have a line sc.close() to close the scanner once you are done scanning.

engenhui1999 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,104 @@
+            }
+        }
+
+        String farewell = duke.farewellUser();
```
I think it is also possible to directly print the message within the method itself, instead of retrieving the string and printing it.

Also, I think if you want to get the farewell message for the user, I believe the naming convention should be "getFarewellUser()" which will be more comprehensible.

engenhui1999 reviewed (commented):
I think it will be good to have more OOP when creating new classes, I think the subsequent tasks required will require you to change quite abit of your code for better OOP principles. Jiayous!!

ZHTang29 committed:
Made changes for more OOP & added save functionality

ZHTang29 committed:
Add Date/Time functionality

ZHTang29 committed:
Added functionality for Date/Time

ZHTang29 committed:
Remove unnecessary import statements

ZHTang29 committed:
Separated classes into packages

ZHTang29 committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

ZHTang29 committed:
Added gradle support

ZHTang29 committed:
Added some JUnit tests

ZHTang29 committed:
Added JavaDocs for all classes/methods

ZHTang29 committed:
Tweaked code to comply with coding standard

ZHTang29 committed:
Added find by keyword functionality

ZHTang29 committed:
Fixed merge conflict issues

ZHTang29 committed:
Fixed merging conflict for new find command

ZHTang29 committed:
Implement GUI

ZHTang29 committed:
Added relevant JavaDocs to (new) methods

There are several methods that are missing JavaDocs. In some cases,
it is hard to tell what the method is doing without the proper
documentation.

Implementing the missing JavaDoc for such code improves the readabiliity
of the code and makes it more understandable to the users.

ZHTang29 committed:
Added assertions to some methods

Assertions help to trigger alarms and inform the user when a bug
appears in their program.

Adding assertions to methods that require them will help us prevent bugs
from occuring in our final version of the product.

ZHTang29 committed:
Refactor some classes into Utilties package

There are certain classes that were previously not classified in any
specfic package. These classes included all the utilities related to
Duke, such as the Storage, UI etc..

Refactoring these classses improves the overall quality of code.

ZHTang29 committed:
Merge pull request #2 from ZHTang29/branch-A-Assertions

Added assertions to some methods

ZHTang29 committed:
Merge branch 'master' into branch-A-CodeQuality

ZHTang29 committed:
Merge pull request #3 from ZHTang29/branch-A-CodeQuality

Refactor some classes into Utilities package

ZHTang29 committed:
Added Notes functionality

The previous version of Duke only allowed Users to store Data in Duke if
the data was in the form of a To Do item, an event or a deadline.

Adding the Notes functionality will enable users to better store data in
Duke.

As a step toward better coding quality, all new methods have JavaDoc
whereever necessary and the new functionality is put in a new package.

ZHTang29 committed:
Add more Error Handling

There were several methods that previously threw unhandled
exceptions/poorly handled exceptions.

Exception handling is especially important in improving the coding
quality especially since Duke runs on a very specific chain of commands
and inputs.

As we work toward Duke with less bugs, we added the error handling for
the 'mark', 'unmark', 'delete', 'deletenote' for situations where the
command was input with non-integer arguments.

Additionally, added error handling for the situation where users try to
add an invalid event (the end date is before the start date).

ZHTang29 committed:
Fix text to box

ZHTang29 committed:
Added User Guide

ZHTang29 committed:
Improve readability of User Guide

ZHTang29 committed:
Renamed product screenshot
